power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,485 @@
+            attributes[element_number_] =
+                &component.get_attribute(attributes_span[element_number_].as<std::string_view>());
+        }
+        element_number_ = -1;
+        return attributes;
+    }
+
+    void count_data() {
+        msgpack::object const& obj = get_value_from_root("data", is_batch_ ? msgpack::type::ARRAY : msgpack::type::MAP);
+        buffers_ = {};
+        // pointer to array (or single value) of msgpack objects to the data
+        ArraySpan batch_data;
+        if (is_batch_) {
+            // NOLINTBEGIN(cppcoreguidelines-pro-type-union-access)
+            batch_size_ = static_cast<Idx>(obj.via.array.size);
+            batch_data = {obj.via.array.ptr, obj.via.array.size};
+            // NOLINTEND(cppcoreguidelines-pro-type-union-access)
+        } else {
+            batch_size_ = 1;
+            batch_data = {&obj, 1};
+        }
+
+        // get set of all components
+        std::set<MetaComponent const*> all_components;
+        for (scenario_number_ = 0; scenario_number_ != static_cast<Idx>(batch_data.size()); ++scenario_number_) {
+            msgpack::object const& scenario = batch_data[scenario_number_];
+            if (scenario.type != msgpack::type::MAP) {
+                throw SerializationError{"The data object of each scenario should be a dictionary!\n"};
+            }
+            for (msgpack::object_kv const& kv : scenario.as<MapSpan>()) {
+                component_key_ = key_to_string(kv);
+                all_components.insert(&dataset_->get_component(component_key_));
+            }
+            component_key_ = "";
+        }
+        scenario_number_ = -1;
+
+        // create buffer object
+        for (MetaComponent const* const component : all_components) {
+            buffers_.push_back(count_component(batch_data, *component));
+        }
+    }
+
+    Buffer count_component(ArraySpan batch_data, MetaComponent const& component) {
+        component_key_ = component.name;
+        // count number of element of all scenarios
+        IdxVector counter(batch_size_);
+        std::vector<ArraySpan> msg_data(batch_size_);
+        for (scenario_number_ = 0; scenario_number_ != batch_size_; ++scenario_number_) {
+            msgpack::object const& scenario = batch_data[scenario_number_];
+            Idx const found_component_idx = find_key_from_map(scenario, component.name);
+            if (found_component_idx >= 0) {
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+                msgpack::object const& element_array = scenario.via.map.ptr[found_component_idx].val;
+                if (element_array.type != msgpack::type::ARRAY) {
+                    throw SerializationError{"Each entry of component per scenario should be a list!\n"};
+                }
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+                counter[scenario_number_] = static_cast<Idx>(element_array.via.array.size);
```

Comment:
optional: move the accessors to separate functions so that the `NOLINT` is only mentioned in standalone locations, e.g. a function that the equivalent of

```cpp
static auto as_array(msgpack::object& element_array) { return element_array.via.array; }
static auto as_map(msgpack::object& element_array) { return element_array.via.map; }
```

or use full functions to get the size and component value